src/btree/bt_walk.c

```diff
@@ -624,3 +625,241 @@ __wti_tree_walk_skip(WT_SESSION_IMPL *session, WT_REF **refp, uint64_t *skipleaf
+ *     suitable page.
+ *
+ * NOTE: Must be called within WT_WITH_PAGE_INDEX or WT_ENTER_PAGE_INDEX
+ */
+static int
+__page_from_npos_internal(WT_SESSION_IMPL *session, WT_REF **refp, uint32_t flags, double npos)
+{
+    WT_BTREE *btree;
+    WT_DECL_RET;
+    WT_PAGE *page;
+    WT_PAGE_INDEX *pindex;
+    WT_REF *current, *descent;
+    double npos_local;
+    int idx, entries;
+    bool eviction;
+
+    /*
+     * The content of this function is derived from the `random_descent` one but has diverged since.
+     */
+    *refp = NULL;
+
+    btree = S2BT(session);
+    current = NULL;
+    /*
+     * This function is called by eviction to find a page in the cache. That case is indicated by
+     * the WT_READ_CACHE flag. Ordinary lookups in a tree will read pages into cache as needed.
+     */
+    eviction = LF_ISSET(WT_READ_CACHE);
+
+    if (0) {
+restart:; /* Restart the search from the root. */
+    }
+
+    /* Search the internal pages of the tree. */
+    current = &btree->root;
+    npos_local = npos;
+    for (;;) {
+        if (F_ISSET(current, WT_REF_FLAG_LEAF))
+            goto done;
+
+        /* The entire for loop's body is for INTERNAL pages only */
+
+        page = current->page;
+        WT_INTL_INDEX_GET(session, page, pindex);
+        entries = (int)pindex->entries;
+
+        npos_local *= entries;
+        idx = (int)npos_local;
+        idx = WT_CLAMP(idx, 0, entries - 1);
+        npos_local -= idx;
+        descent = pindex->index[idx];
+
+        if (eviction) {
+            /*
+             * In case of eviction, we never want to load pages from disk. Also, page_swap with
+             * WT_READ_CACHE will fail anyway and we'll lose our pointer, so avoid making a call
+             * that will fail.
+             */
+            switch (WT_REF_GET_STATE(descent)) {
+            case WT_REF_DISK:
+            case WT_REF_LOCKED:
+            case WT_REF_DELETED:
+                /* Can't go down from here but it's ok to return this page */
+                goto done;
+            default: /* WT_REF_MEM, WT_REF_SPLIT */
+                goto descend;
+            }
+            /* Unreachable */
+        } else {
+            /* Not eviction */
+            switch (WT_REF_GET_STATE(descent)) {
+            case WT_REF_LOCKED:
+                if (!LF_ISSET(WT_READ_NO_WAIT)) {
+                    WT_RET(__wt_page_release(session, current, flags));
+                    __wt_sleep(0, 10);
+                    goto restart;
+                }
+                /* Fall through */
+            case WT_REF_DELETED:
+                /*
+                 * Can't go down from here. Return this page and __find_closest_page will finish the
+                 * job.
+                 */
+                goto done;
+            default: /* WT_REF_DISK, WT_REF_MEM, WT_REF_SPLIT */
+                goto descend;
+            }
+            /* Unreachable */
+        }
+        /* Unreachable */
+
+descend:
```

Comment:
Code after an unreachable block is a code smell. Can we have a tech debt ticket to clean this up post-PR?

Reply:
There's a goto label implies goto's. Not sure how it might be ambiguous.